File: 1277/1277_Count_Square_Submatrices_with_All_Ones.py
class Solution(object):
    def countSquares(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        h = len(matrix)
        w = len(matrix[0])
        self.res = []
        for i in xrange(h):
            self.res.append([])
        res = 0
        for i in xrange(h):
            for j in xrange(w):
                size_max_square = self.findMaxSquare(matrix, i, j)
                # print i, j, size_max_square
                self.res[i].append(size_max_square)
                res += size_max_square
        return res

    def testNextSquare(self, matrix, i, j, size):
        if matrix[i+size-1][j+size-1] == 0:
            return False
        for delta in xrange(size):
            if matrix[i+delta][j+size-1] == 0 or matrix[i+size-1][j+delta] == 0:
                return False
        return True

    def maxPossibleSquare(self, matrix, i, j):
        h = len(matrix)
        w = len(matrix[0])
        return min(h-i, w-j)

    def findMaxSquare(self, matrix, i ,j):
        if matrix[i][j] == 0:
            return 0
        if i > 0 and j > 0:
            res_topleft = self.res[i-1][j-1]
            size_max_square = max(1, res_topleft-1)
        else:
            size_max_square = 1

        for size in xrange(size_max_square+1, self.maxPossibleSquare(matrix, i, j)+1):
            if not self.testNextSquare(matrix, i ,j, size):
                break
            size_max_square = size

        return size_max_square


test = Solution()
print test.countSquares(
    [
      [0,1,1,1],
      [1,1,1,1],
      [0,1,1,1]
    ]
) # 15

print test.countSquares(
    [
      [1,0,1],
      [1,1,0],
      [1,1,0]
    ]
) # 7

File: 1277/1277_Count_Square_Submatrices_with_All_Ones.java
import java.util.Arrays;

class Solution {
    public int countSquares(int[][] matrix) {
        if (matrix.length == 0) return 0;
        // Stores number of all-one square submatrices with top-left at memoi[r][c]
        int[][] memoi = new int[matrix.length][matrix[0].length];
        for (int i = 0; i < memoi.length; i++) {
            Arrays.fill(memoi[i], -1);
        }
        int res = 0;
        for (int i = 0; i < memoi.length; i++) {
            for (int j = 0; j < memoi[i].length; j++) {
                res += dp(i, j, matrix, memoi);
            }
        }
        return res;
    }
    int dp(int r, int c, int[][] matrix, int[][]memoi) {
        // Handle out of bound
        if (r >= matrix.length || c >= matrix[0].length) {
            return 0;
        }
        if (memoi[r][c] >= 0) return memoi[r][c];
        if (matrix[r][c] == 0) {
            // If top-left is 0, there is no all-one submatrix starting here
            memoi[r][c] = 0;
            return 0;
        }
        // Out of bound will be handled within recursive call
        int down = dp(r+1, c, matrix, memoi);
        int right = dp(r, c+1, matrix, memoi);
        int diag = dp(r+1, c+1, matrix, memoi);
        // If matrix[r][c] is 1, then we have at least 1,
        // plus the minimum of results of down, right and diag.
        // Note the matrix has to be square,
        // So there is a 2x2 all-one iff all 3 results are 1 (all single-element matrix),
        // and so on so forth
        int res = 1 + Math.min(down, Math.min(right, diag));
        memoi[r][c] = res;
        return res;
    }
}
